File: ctb/sys/libs/orbt-utils/heir-generator.js
/**
 * HEIR/ORBT ID Generator
 * Version: 1.0.0
 * Date: 2025-10-23
 * 
 * Generates Barton Doctrine compliant HEIR and Process IDs
 * for tracking operations across the ORBT (Operational Resource Behavior Tracking) system
 * 
 * HEIR ID Format: HEIR-YYYY-MM-SYSTEM-MODE-VN
 * Process ID Format: PRC-SYSTEM-EPOCHTIMESTAMP
 * 
 * ORBT Layers:
 * - Layer 1: Input/Intake operations
 * - Layer 2: Processing/Middle operations
 * - Layer 3: Output/Generation operations
 * - Layer 4: Orchestration/Coordination operations
 */

class HeirGenerator {
    /**
     * Generate a unique HEIR ID
     * @param {string} system - System identifier (e.g., 'IMO', 'ACTION', 'INTAKE')
     * @param {string} mode - Operation mode (e.g., 'CREATOR', 'ACTION', 'VALIDATE')
     * @param {number} version - Version number (default: 1)
     * @returns {string} HEIR ID in format HEIR-YYYY-MM-SYSTEM-MODE-VN
     */
    static generateHeirId(system, mode, version = 1) {
        const now = new Date();
        const year = now.getFullYear();
        const month = String(now.getMonth() + 1).padStart(2, '0');
        const versionPadded = String(version).padStart(2, '0');
        
        return `HEIR-${year}-${month}-${system}-${mode}-${versionPadded}`;
    }

    /**
     * Generate a unique Process ID
     * @param {string} system - System identifier (e.g., 'IMO', 'ACTION', 'INTAKE')
     * @returns {string} Process ID in format PRC-SYSTEM-EPOCHTIMESTAMP
     */
    static generateProcessId(system) {
        const timestamp = Date.now();
        return `PRC-${system}-${timestamp}`;
    }

    /**
     * Generate a complete HEIR/ORBT payload
     * @param {string} tool - Tool name (e.g., 'apify_run_actor', 'gmail_send')
     * @param {object} data - Tool-specific data payload
     * @param {string} system - System identifier
     * @param {string} mode - Operation mode
     * @param {number} orbtLayer - ORBT layer (1-4)
     * @param {string} blueprintVersion - Blueprint version (default: '1.0')
     * @param {number} version - HEIR version number (default: 1)
     * @returns {object} Complete HEIR/ORBT payload
     */
    static generatePayload(tool, data, system, mode, orbtLayer, blueprintVersion = '1.0', version = 1) {
        // Validate ORBT layer
        if (orbtLayer < 1 || orbtLayer > 4) {
            throw new Error(`Invalid ORBT layer: ${orbtLayer}. Must be between 1 and 4.`);
        }

        return {
            tool: tool,
            data: data,
            unique_id: this.generateHeirId(system, mode, version),
            process_id: this.generateProcessId(system),
            orbt_layer: orbtLayer,
            blueprint_version: blueprintVersion
        };
    }

    /**
     * Parse a HEIR ID into its components
     * @param {string} heirId - HEIR ID to parse
     * @returns {object} Parsed components
     */
    static parseHeirId(heirId) {
        const pattern = /^HEIR-(\d{4})-(\d{2})-([A-Z]+)-([A-Z]+)-(\d{2})$/;
        const match = heirId.match(pattern);

        if (!match) {
            throw new Error(`Invalid HEIR ID format: ${heirId}`);
        }

        return {
            year: match[1],
            month: match[2],
            system: match[3],
            mode: match[4],
            version: parseInt(match[5], 10)
        };
    }

    /**
     * Parse a Process ID into its components
     * @param {string} processId - Process ID to parse
     * @returns {object} Parsed components
     */
    static parseProcessId(processId) {
        const pattern = /^PRC-([A-Z]+)-(\d+)$/;
        const match = processId.match(pattern);

        if (!match) {
            throw new Error(`Invalid Process ID format: ${processId}`);
        }

        return {
            system: match[1],
            timestamp: parseInt(match[2], 10),
            date: new Date(parseInt(match[2], 10))
        };
    }

    /**
     * Validate a HEIR/ORBT payload
     * @param {object} payload - Payload to validate
     * @returns {boolean} True if valid, throws error otherwise
     */
    static validatePayload(payload) {
        const required = ['tool', 'data', 'unique_id', 'process_id', 'orbt_layer', 'blueprint_version'];
        
        for (const field of required) {
            if (!(field in payload)) {
                throw new Error(`Missing required field: ${field}`);
            }
        }

        // Validate HEIR ID format
        this.parseHeirId(payload.unique_id);

        // Validate Process ID format
        this.parseProcessId(payload.process_id);

        // Validate ORBT layer
        if (payload.orbt_layer < 1 || payload.orbt_layer > 4) {
            throw new Error(`Invalid ORBT layer: ${payload.orbt_layer}`);
        }

        return true;
    }

    /**
     * Get ORBT layer description
     * @param {number} layer - ORBT layer number (1-4)
     * @returns {string} Layer description
     */
    static getOrbtLayerDescription(layer) {
        const descriptions = {
            1: 'Input/Intake - Data ingestion and validation',
            2: 'Processing/Middle - Data transformation and enrichment',
            3: 'Output/Generation - Result creation and formatting',
            4: 'Orchestration/Coordination - System-level coordination'
        };

        return descriptions[layer] || 'Unknown layer';
    }
}

// Export for Node.js
if (typeof module !== 'undefined' && module.exports) {
    module.exports = HeirGenerator;
}

// Example usage:
/*
const HeirGenerator = require('./heir-generator');

// Generate a simple HEIR ID
const heirId = HeirGenerator.generateHeirId('IMO', 'CREATOR', 1);
console.log(heirId); // HEIR-2025-10-IMO-CREATOR-01

// Generate a Process ID
const processId = HeirGenerator.generateProcessId('IMO');
console.log(processId); // PRC-IMO-1729651200000

// Generate a complete payload
const payload = HeirGenerator.generatePayload(
    'apify_run_actor',
    { actorId: 'apify~leads-finder', runInput: { query: 'test' } },
    'IMO',
    'ACTION',
    2,
    '1.0',
    1
);
console.log(JSON.stringify(payload, null, 2));

// Validate payload
try {
    HeirGenerator.validatePayload(payload);
    console.log('✅ Payload is valid');
} catch (error) {
    console.error('❌ Payload validation failed:', error.message);
}
*/

File: ctb/ai/orbt-utils/heir_generator.py
# # CTB Metadata
# # Generated: 2025-10-23T14:32:41.010631
# # CTB Version: 1.3.3
# # Division: AI Agents & MCP
# # Category: orbt-utils
# # Compliance: 100%
# # HEIR ID: HEIR-2025-10-AI-ORBT-U-01

"""
HEIR/ORBT ID Generator
Version: 1.0.0
Date: 2025-10-23

Generates Barton Doctrine compliant HEIR and Process IDs
for tracking operations across the ORBT (Operational Resource Behavior Tracking) system

HEIR ID Format: HEIR-YYYY-MM-SYSTEM-MODE-VN
Process ID Format: PRC-SYSTEM-EPOCHTIMESTAMP

ORBT Layers:
- Layer 1: Input/Intake operations
- Layer 2: Processing/Middle operations
- Layer 3: Output/Generation operations
- Layer 4: Orchestration/Coordination operations
"""

import re
from datetime import datetime
from typing import Dict, Any


class HeirGenerator:
    """Generator for HEIR and Process IDs compliant with Barton Doctrine"""

    @staticmethod
    def generate_heir_id(system: str, mode: str, version: int = 1) -> str:
        """
        Generate a unique HEIR ID
        
        Args:
            system: System identifier (e.g., 'IMO', 'ACTION', 'INTAKE')
            mode: Operation mode (e.g., 'CREATOR', 'ACTION', 'VALIDATE')
            version: Version number (default: 1)
            
        Returns:
            HEIR ID in format HEIR-YYYY-MM-SYSTEM-MODE-VN
        """
        now = datetime.now()
        year = now.year
        month = f"{now.month:02d}"
        version_padded = f"{version:02d}"
        
        return f"HEIR-{year}-{month}-{system}-{mode}-{version_padded}"

    @staticmethod
    def generate_process_id(system: str) -> str:
        """
        Generate a unique Process ID
        
        Args:
            system: System identifier (e.g., 'IMO', 'ACTION', 'INTAKE')
            
        Returns:
            Process ID in format PRC-SYSTEM-EPOCHTIMESTAMP
        """
        timestamp = int(datetime.now().timestamp() * 1000)
        return f"PRC-{system}-{timestamp}"

    @staticmethod
    def generate_payload(
        tool: str,
        data: Dict[str, Any],
        system: str,
        mode: str,
        orbt_layer: int,
        blueprint_version: str = "1.0",
        version: int = 1
    ) -> Dict[str, Any]:
        """
        Generate a complete HEIR/ORBT payload
        
        Args:
            tool: Tool name (e.g., 'apify_run_actor', 'gmail_send')
            data: Tool-specific data payload
            system: System identifier
            mode: Operation mode
            orbt_layer: ORBT layer (1-4)
            blueprint_version: Blueprint version (default: '1.0')
            version: HEIR version number (default: 1)
            
        Returns:
            Complete HEIR/ORBT payload
            
        Raises:
            ValueError: If ORBT layer is not between 1 and 4
        """
        # Validate ORBT layer
        if orbt_layer < 1 or orbt_layer > 4:
            raise ValueError(f"Invalid ORBT layer: {orbt_layer}. Must be between 1 and 4.")

        return {
            "tool": tool,
            "data": data,
            "unique_id": HeirGenerator.generate_heir_id(system, mode, version),
            "process_id": HeirGenerator.generate_process_id(system),
            "orbt_layer": orbt_layer,
            "blueprint_version": blueprint_version
        }

    @staticmethod
    def parse_heir_id(heir_id: str) -> Dict[str, Any]:
        """
        Parse a HEIR ID into its components
        
        Args:
            heir_id: HEIR ID to parse
            
        Returns:
            Dictionary with parsed components
            
        Raises:
            ValueError: If HEIR ID format is invalid
        """
        pattern = r'^HEIR-(\d{4})-(\d{2})-([A-Z]+)-([A-Z]+)-(\d{2})$'
        match = re.match(pattern, heir_id)

        if not match:
            raise ValueError(f"Invalid HEIR ID format: {heir_id}")

        return {
            "year": match.group(1),
            "month": match.group(2),
            "system": match.group(3),
            "mode": match.group(4),
            "version": int(match.group(5))
        }

    @staticmethod
    def parse_process_id(process_id: str) -> Dict[str, Any]:
        """
        Parse a Process ID into its components
        
        Args:
            process_id: Process ID to parse
            
        Returns:
            Dictionary with parsed components
            
        Raises:
            ValueError: If Process ID format is invalid
        """
        pattern = r'^PRC-([A-Z]+)-(\d+)$'
        match = re.match(pattern, process_id)

        if not match:
            raise ValueError(f"Invalid Process ID format: {process_id}")

        timestamp = int(match.group(2))
        return {
            "system": match.group(1),
            "timestamp": timestamp,
            "date": datetime.fromtimestamp(timestamp / 1000)
        }

    @staticmethod
    def validate_payload(payload: Dict[str, Any]) -> bool:
        """
        Validate a HEIR/ORBT payload
        
        Args:
            payload: Payload to validate
            
        Returns:
            True if valid
            
        Raises:
            ValueError: If payload is invalid
        """
        required = ['tool', 'data', 'unique_id', 'process_id', 'orbt_layer', 'blueprint_version']
        
        for field in required:
            if field not in payload:
                raise ValueError(f"Missing required field: {field}")

        # Validate HEIR ID format
        HeirGenerator.parse_heir_id(payload['unique_id'])

        # Validate Process ID format
        HeirGenerator.parse_process_id(payload['process_id'])

        # Validate ORBT layer
        if payload['orbt_layer'] < 1 or payload['orbt_layer'] > 4:
            raise ValueError(f"Invalid ORBT layer: {payload['orbt_layer']}")

        return True

    @staticmethod
    def get_orbt_layer_description(layer: int) -> str:
        """
        Get ORBT layer description
        
        Args:
            layer: ORBT layer number (1-4)
            
        Returns:
            Layer description
        """
        descriptions = {
            1: 'Input/Intake - Data ingestion and validation',
            2: 'Processing/Middle - Data transformation and enrichment',
            3: 'Output/Generation - Result creation and formatting',
            4: 'Orchestration/Coordination - System-level coordination'
        }

        return descriptions.get(layer, 'Unknown layer')


# Example usage:
if __name__ == "__main__":
    # Generate a simple HEIR ID
    heir_id = HeirGenerator.generate_heir_id('IMO', 'CREATOR', 1)
    print(f"HEIR ID: {heir_id}")

    # Generate a Process ID
    process_id = HeirGenerator.generate_process_id('IMO')
    print(f"Process ID: {process_id}")

    # Generate a complete payload
    payload = HeirGenerator.generate_payload(
        'apify_run_actor',
        {'actorId': 'apify~leads-finder', 'runInput': {'query': 'test'}},
        'IMO',
        'ACTION',
        2,
        '1.0',
        1
    )
    print(f"\nPayload:\n{payload}")

    # Validate payload
    try:
        HeirGenerator.validate_payload(payload)
        print("\n✅ Payload is valid")
    except ValueError as error:
        print(f"\n❌ Payload validation failed: {error}")

    # Parse IDs
    parsed_heir = HeirGenerator.parse_heir_id(heir_id)
    print(f"\nParsed HEIR ID: {parsed_heir}")

    parsed_process = HeirGenerator.parse_process_id(process_id)
    print(f"Parsed Process ID: {parsed_process}")

    # Get layer description
    for layer in range(1, 5):
        desc = HeirGenerator.get_orbt_layer_description(layer)
        print(f"Layer {layer}: {desc}")
